### PR TITLE
#92 Create function to derive reference dates in DM domain

### LINKED ISSUE #92: Create function to derive the reference dates
### Background Information

Derivation function to derive the reference min dates and maximum dates which can be used in the DM domain for deriving reference start and end dates.

### Definition of Done

_No response_

### PR DESCRIPTION
**Pull request by muzzama-1990**
Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/sdtm.oak/CONTRIBUTING.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #92 into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/oak_cal_ref_dates.R [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                              min_max = "min",
+                              ref_date_config_df,
+                              raw_source) {
+  # Check if ref_date_config_df is a data frame and has all required variables
+  admiraldev::assert_data_frame(ref_date_config_df, required_vars = exprs(
+    dataset_name, date_var,
+    time_var, dformat,
+    tformat, sdtm_var_name
+  ))
+
```
Comment (ShiyuC) [inline R/oak_cal_ref_dates.R:78]: Should we add some assertion to `raw_source`? For example `raw_source` should be a list and length(raw_source) > 0.
Also, the "dataset_name" mentioned in `ref_date_config_df` should exist in `raw_source`

File: R/cal_min_max_date.R [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#'   time_format = "H:M"
+#' )
+#'
+cal_min_max_date <- function(raw_dataset,
+                             date_variable,
+                             time_variable,
+                             val_type = "min",
+                             date_format,
+                             time_format) {
+  # Check if date is present in the raw data frame
```
Comment (ShiyuC) [inline R/cal_min_max_date.R:50]: I think we need to check if `date_variable`, `time_variable` are a string of single character, in case users input a vector of variable names.

**Review by ShiyuC — COMMENTED** [review]
@muzzama-1990 Thanks for your great work on these functions! I left some comments for your consideration.

File: R/cal_min_max_date.R [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    cli::cli_warn(paste(
+      "Date variable", date_variable, "or Time variable", time_variable,
+      "not present in source data"
+    ))
+    return(empty_df)
+  }
+
+  fin_df <- raw_dataset
+  # Time variable is not used then use only date
+  if (is.na(time_variable)) {
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:72]: I am indented to check if the time variable is not used and is empty string. This has been updated to length()==0.

File: R/cal_min_max_date.R [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+#' )
+#'
+cal_min_max_date <- function(raw_dataset,
+                             date_variable,
+                             time_variable,
+                             val_type = "min",
+                             date_format,
+                             time_format) {
+  # Check if date is present in the raw data frame
+  date_not_in_data <- !(date_variable %in% colnames(raw_dataset))
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:51]: updated accordingly

File: R/cal_min_max_date.R [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+  # Check if time variable is used and present in the raw data frame
+  time_not_in_data <- !(time_variable %in% colnames(raw_dataset)) && !is.na(time_variable)
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:54]: Updated

File: R/cal_min_max_date.R [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  fin_df <- raw_dataset
+  # Time variable is not used then use only date
+  if (is.na(time_variable)) {
+    fin_df$datetime <- create_iso8601(raw_dataset[[date_variable]],
+      .format = date_format
+    )
+  } else {
+    # If both date and time variables are presen use both date and time
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:77]: updated

File: R/cal_min_max_date.R [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    return(empty_df)
+  }
+
+  fin_df <- raw_dataset
+  # Time variable is not used then use only date
+  if (is.na(time_variable)) {
+    fin_df$datetime <- create_iso8601(raw_dataset[[date_variable]],
+      .format = date_format
+    )
+  } else {
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:76]: If date parameter is missing or NULL then the function returns the empty data frame as we cannot calculate the min/max date if date parameter is missing.
Test case is added to check this scenario.

File: R/cal_min_max_date.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/cal_min_max_date.R:72]: I am checking if the time_variable parameter is not passed and is missing.

File: R/oak_cal_ref_dates.R [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#' @description Derive RFSTDTC, RFENDTC, RFXENDTC, RFXSTDTC based on the input dates and time.
+#'
+#'
+#' @param ds_in Data frame. DM domain.
+#' @param der_var Character string. The reference date to be derived.
+#' @param min_max Minimum or Maximum date to be calculated based on the input.
+#'        Default set to Minimum. Values should be min or max.
+#' @param ref_date_config_df Data frame which has the details of the variables to
+#'        be used for the calculation of reference dates.
+#'        Should has columns listed below:
```
[the remark this answers is not in the document]
Comment (muzzama-1990) [reply R/oak_cal_ref_dates.R:14]: updated

File: R/oak_cal_ref_dates.R (hunk 1 above)
Comment (muzzama-1990) [reply R/oak_cal_ref_dates.R:78]: Added assertion to check if the parameter passed is a list and also include a warning message when the data frame referenced in the ref_date_config_df is not present in the input raw_source

**Review by ShiyuC — APPROVED** [review]
Looks good to me, thanks Mohsin!

**Review by rammprasad — APPROVED** [review]
Looks good to me. Made minor changes to the examples/code. Amazing work, Mohsin.
